lxd/api.go

```diff
@@ -208,12 +213,76 @@ func restServer(d *Daemon) *http.Server {
 		_ = response.NotFound(nil).Render(w)
 	})
 
+	// Initialize API metrics and use middleware functions to update API request metrics.
+	metrics.InitAPIMetrics()
+	mux.Use(func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			metrics.MeasureOngoingRequest(*r.URL)
+			next.ServeHTTP(w, r)
+		})
+	})
+
+	mux.Use(func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			wrapper := &responseWriterWrapper{writer: w}
+			next.ServeHTTP(wrapper, r)
+
+			if !wrapper.AsyncResponse {
+				metrics.MeasureCompletedRequest(*r.URL, http.StatusOK, false)
+			}
+		})
+	}) // This has to be the last middleware function to be defined/called.
+
 	return &http.Server{
 		Handler:     &lxdHTTPServer{r: mux, d: d},
 		ConnContext: lxdRequest.SaveConnectionInContext,
 	}
 }
 
+type responseWriterWrapper struct {
+	writer        http.ResponseWriter
+	StatusCode    int
+	AsyncResponse bool
+}
+
+// WriteHeader calls the underlying WriteHeader and catches the response status code for future use.
+func (rw *responseWriterWrapper) WriteHeader(statusCode int) {
+	rw.StatusCode = statusCode
+	rw.writer.WriteHeader(statusCode)
+}
+
+// Header asserts the wrapper implements http.ResponseWriter.
+func (rw *responseWriterWrapper) Header() http.Header {
+	return rw.writer.Header()
+}
+
+// Write calls the underlying Write and checks from the response body if the status is api.OperationResponse.
+func (rw *responseWriterWrapper) Write(body []byte) (int, error) {
+	reader := bytes.NewReader(body)
+	decoder := json.NewDecoder(reader)
+	response := api.Response{}
+
+	// Catch and decode the response to extract the status code from the body.
+	err := decoder.Decode(&response)
+	if err == nil {
+		// Check the status code for operation creation.
+		if response.StatusCode == int(api.OperationCreated) {
+			rw.AsyncResponse = true
+		}
+	}
+	return rw.writer.Write(body)
+}
+
+// Hijack asserts the wrapper implements http.Hijack interface for websocket upgrade.
+func (rw *responseWriterWrapper) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	h, ok := rw.writer.(http.Hijacker)
+	if ok {
+		return h.Hijack()
+	}
+
+	return nil, nil, fmt.Errorf("Inner wrapper does not implement http.Hijacker")
+}
```

Comment:
This log line makes me suspicious: https://github.com/canonical/lxd/actions/runs/10161658665/job/28100619559#step:12:1396

Reply:
I am currently trying to figure out what is this hanhing process, but I hadn't thought about Hijack, thanks for the insight!

Reply:
LXD was failing to shutdown because the wrapper did not implement `http.Flusher`, see I believe [this](https://github.com/hamistao/lxd/blob/c0cfe23fe79614f658456c2593e1fa07974957a1/lxd/api_internal.go#L256) is the critical point of failure. Should be OK now.